Allow user to manually pass module.name associated with global in {add}_safe_global

mikaylagawarecki opened the pull request:
Fixes #142144

A global x is saved in checkpoint as `GLOBAL x.__module__ x.__name__`. So , after allowlisting a GLOBAL it is expected to match any GLOBAL instruction of the form `GLOBAL x.__module__ x.__name__`  but there are edge cases when for the same API from the same module, what `__module__` gives changes between versions which prevents users from allowlisting the global.

In this case, in numpy < 2.1

```
torch.save("bla", np_array)
# checkpoint has GLOBAL "np.core.multiarray" "_reconstruct"
```
In np version 2.1 

```
with safe_globals([np.core.multiarray._reconstruct]):
    torch.load("bla")
```
np.core.multiarray._reconstruct.__module__ gives "np._core.multiarray" (note the extra _ before core) and see what was done [here](https://github.com/numpy/numpy/blob/main/numpy/core/multiarray.py)

Since the dictionary to access safe globals is keyed on "{foo.__module__}.{foo.__name__}", __module__, __name__ will no longer match that in the checkpoint so "np.core.multiarray._reconstruct" can no longer be properly allowlisted (instead np._core.multiarray._reconstruct is a key in the dict).

We allow `add_safe_globals/safe_globals` to optionally take tuples of (global, str of module.name) to workaround such (odd/edge case) situations.



Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #142153

mikaylagawarecki committed:
Allow user to manually pass module.name associated with global in {add}_safe_global

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Allow user to manually pass module.name associated with global in {add}_safe_global"

Fixes #142144

A global x is saved in checkpoint as `GLOBAL x.__module__ x.__name__`. So , after allowlisting a GLOBAL it is expected to match any GLOBAL instruction of the form `GLOBAL x.__module__ x.__name__`  but there are edge cases when for the same API from the same module, what `__module__` gives changes between versions which prevents users from allowlisting the global.

In this case, in numpy < 2.1

```
torch.save("bla", np_array)
# checkpoint has GLOBAL "np.core.multiarray" "_reconstruct"
```
In np version 2.1 

```
with safe_globals([np.core.multiarray._reconstruct]):
    torch.load("bla")
```
np.core.multiarray._reconstruct.__module__ gives "np._core.multiarray" (note the extra _ before core) and see what was done [here](https://github.com/numpy/numpy/blob/main/numpy/core/multiarray.py)

Since the dictionary to access safe globals is keyed on "{foo.__module__}.{foo.__name__}", __module__, __name__ will no longer match that in the checkpoint so "np.core.multiarray._reconstruct" can no longer be properly allowlisted (instead np._core.multiarray._reconstruct is a key in the dict).

We allow `add_safe_globals/safe_globals` to optionally take tuples of (global, str of module.name) to workaround such (odd/edge case) situations.






[ghstack-poisoned]

albanD commented on torch/_weights_only_unpickler.py:
```diff
@@ -127,8 +127,14 @@ def __exit__(self, type, value, tb):
-        rc[f"{module}.{name}"] = f
+        if isinstance(f, tuple):
+            assert len(f) == 2, "Expected tuple of (global, str of module.name)"
+            assert type(f[1]) is str, "Expected str of module.name"
```
These are ValueError and TypeError, not assertion errors.

albanD commented on torch/serialization.py:
```diff
@@ -274,6 +274,17 @@ def add_safe_globals(safe_globals: List[Any]) -> None:
     added to this list can be called during unpickling, classes could be instantiated
     and have state set.
 
+    Each item, ``i`` in the list can either be a function/class or a tuple of the form
```
Let's change the `safe_globals dtype to List[Callable | Tuple[Callable, str]]

albanD commented on torch/serialization.py:
```diff
@@ -274,6 +274,17 @@ def add_safe_globals(safe_globals: List[Any]) -> None:
+    (function/class, string) where string is a string representing ``{__module__}.{__name__}``
+    of the function/class.
+
+    If a single item ``i`` is passed, ``i`` will be the associated function/class for
```
Within the serialized format, each function is identified with it's full path as ``{__module__}.{__name__}``. When calling this API, you can provide this full path that should match the one in the checkpoint otherwise the default `{fn.__module__}.{fn.__name__}` will be used.

mikaylagawarecki committed:
Update on "Allow user to manually pass module.name associated with global in {add}_safe_global"

Fixes #142144

A global x is saved in checkpoint as `GLOBAL x.__module__ x.__name__`. So , after allowlisting a GLOBAL it is expected to match any GLOBAL instruction of the form `GLOBAL x.__module__ x.__name__`  but there are edge cases when for the same API from the same module, what `__module__` gives changes between versions which prevents users from allowlisting the global.

In this case, in numpy < 2.1

```
torch.save("bla", np_array)
# checkpoint has GLOBAL "np.core.multiarray" "_reconstruct"
```
In np version 2.1 

```
with safe_globals([np.core.multiarray._reconstruct]):
    torch.load("bla")
```
np.core.multiarray._reconstruct.__module__ gives "np._core.multiarray" (note the extra _ before core) and see what was done [here](https://github.com/numpy/numpy/blob/main/numpy/core/multiarray.py)

Since the dictionary to access safe globals is keyed on "{foo.__module__}.{foo.__name__}", __module__, __name__ will no longer match that in the checkpoint so "np.core.multiarray._reconstruct" can no longer be properly allowlisted (instead np._core.multiarray._reconstruct is a key in the dict).

We allow `add_safe_globals/safe_globals` to optionally take tuples of (global, str of module.name) to workaround such (odd/edge case) situations.






[ghstack-poisoned]

albanD commented on torch/_weights_only_unpickler.py:
```diff
@@ -127,8 +129,20 @@ def __exit__(self, type, value, tb):
+                )
+            if type(f[1]) is not str:
+                raise TypeError(
+                    f"Expected second item in tuple to be str of module.name, but got: {type(f[1])}"
```
module.name -> function full path?

albanD commented on torch/serialization.py:
```diff
@@ -261,21 +261,29 @@ def clear_safe_globals() -> None:
     added to this list can be called during unpickling, classes could be instantiated
     and have state set.
 
+    Each item, ``i`` in the list can either be a function/class or a tuple of the form
```
```suggestion
    Each item in the list can either be a function/class or a tuple of the form
```
nit, I don't think it adds much

albanD commented on test/test_cpp_extensions_open_device_registration.py:
```diff
@@ -599,9 +595,16 @@ def test_open_device_numpy_serialization(self):
         with safe_globals(
             [
-                np.core.multiarray._reconstruct,
+                np_reconstruct()
```
This is a good example, but we also need a test to make sure the newly added API works right?

albanD replied:
Ho you already removed it :P

mikaylagawarecki committed:
Update on "Allow user to manually pass module.name associated with global in {add}_safe_global"

Fixes #142144

A global x is saved in checkpoint as `GLOBAL x.__module__ x.__name__`. So , after allowlisting a GLOBAL it is expected to match any GLOBAL instruction of the form `GLOBAL x.__module__ x.__name__`  but there are edge cases when for the same API from the same module, what `__module__` gives changes between versions which prevents users from allowlisting the global.

In this case, in numpy < 2.1

```
torch.save("bla", np_array)
# checkpoint has GLOBAL "np.core.multiarray" "_reconstruct"
```
In np version 2.1 

```
with safe_globals([np.core.multiarray._reconstruct]):
    torch.load("bla")
```
np.core.multiarray._reconstruct.__module__ gives "np._core.multiarray" (note the extra _ before core) and see what was done [here](https://github.com/numpy/numpy/blob/main/numpy/core/multiarray.py)

Since the dictionary to access safe globals is keyed on "{foo.__module__}.{foo.__name__}", __module__, __name__ will no longer match that in the checkpoint so "np.core.multiarray._reconstruct" can no longer be properly allowlisted (instead np._core.multiarray._reconstruct is a key in the dict).

We allow `add_safe_globals/safe_globals` to optionally take tuples of (global, str of module.name) to workaround such (odd/edge case) situations.






[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Allow user to manually pass module.name associated with global in {add}_safe_global"

Fixes #142144

A global x is saved in checkpoint as `GLOBAL x.__module__ x.__name__`. So , after allowlisting a GLOBAL it is expected to match any GLOBAL instruction of the form `GLOBAL x.__module__ x.__name__`  but there are edge cases when for the same API from the same module, what `__module__` gives changes between versions which prevents users from allowlisting the global.

In this case, in numpy < 2.1

```
torch.save("bla", np_array)
# checkpoint has GLOBAL "np.core.multiarray" "_reconstruct"
```
In np version 2.1 

```
with safe_globals([np.core.multiarray._reconstruct]):
    torch.load("bla")
```
np.core.multiarray._reconstruct.__module__ gives "np._core.multiarray" (note the extra _ before core) and see what was done [here](https://github.com/numpy/numpy/blob/main/numpy/core/multiarray.py)

Since the dictionary to access safe globals is keyed on "{foo.__module__}.{foo.__name__}", __module__, __name__ will no longer match that in the checkpoint so "np.core.multiarray._reconstruct" can no longer be properly allowlisted (instead np._core.multiarray._reconstruct is a key in the dict).

We allow `add_safe_globals/safe_globals` to optionally take tuples of (global, str of module.name) to workaround such (odd/edge case) situations.






[ghstack-poisoned]

albanD reviewed (approved):
SGTM!

mikaylagawarecki commented:
@pytorchbot merge
